### PR TITLE
lightning: reuse dupDetector to implement in-memory ingest data

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: ref #45719

Problem Summary:

### What is changed and how it works?

to reuse local backend's logic, we need to let external engine implement IngestData interface. We can simple do it by loading KV pairs from external engine into memory. And to implement the duplicate detection feature, I extract `DupDetector` to let local engine and external engine use same code.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

master

```
goos: darwin
goarch: arm64
pkg: github.com/pingcap/tidb/br/pkg/lightning/backend/local
BenchmarkDupDetectIter
BenchmarkDupDetectIter-8   	       4	 310411136 ns/op
```

this branch

```
goos: darwin
goarch: arm64
pkg: github.com/pingcap/tidb/br/pkg/lightning/backend/local
BenchmarkDupDetectIter
BenchmarkDupDetectIter-8   	       4	 317531698 ns/op
```

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/backend/external/engine.go [hunk 1]
```diff
@@ -58,3 +63,200 @@ func (e *Engine) createMergeIter(ctx context.Context, start kv.Key) (*MergeKVIte
...
+		}
+		lastKeyIdx = len(m.keys) - 1 - i
+	}
+	return firstKeyIdx, lastKeyIdx
+}
+
+// GetFirstAndLastKey implements IngestData.GetFirstAndLastKey.
+func (m *MemoryIngestData) GetFirstAndLastKey(lowerBound, upperBound []byte) ([]byte, []byte, error) {
+	firstKeyIdx, lastKeyIdx := m.firstAndLastKeyIndex(lowerBound, upperBound)
+	if firstKeyIdx < 0 || firstKeyIdx > lastKeyIdx {
```
Comment (tangenta) [inline br/pkg/lightning/backend/external/engine.go:112]: Maybe return an error when `firstKeyIdx > lastKeyIdx` (the keys in `MemoryIngestData` are not sorted)?

File: br/pkg/lightning/backend/external/engine.go [hunk 2]
```diff
@@ -58,3 +63,200 @@ func (e *Engine) createMergeIter(ctx context.Context, start kv.Key) (*MergeKVIte
...
+
+// GetTS implements IngestData.GetTS.
+func (m *MemoryIngestData) GetTS() uint64 {
+	return m.ts
+}
+
+// Finish implements IngestData.Finish.
+func (m *MemoryIngestData) Finish(totalBytes, totalCount int64) {
+	//TODO implement me
+	panic("implement me")
```
Comment (tangenta) [inline br/pkg/lightning/backend/external/engine.go:261]: Will this be modified in future PR?

**Review by tangenta — COMMENTED** [review]
Rest LGTM

File: br/pkg/lightning/backend/external/engine.go (hunk 2 above)
Comment (lance6716) [reply br/pkg/lightning/backend/external/engine.go:261]: yes in next PR

File: br/pkg/lightning/backend/external/engine.go (hunk 1 above)
Comment (lance6716) [reply br/pkg/lightning/backend/external/engine.go:112]: in firstAndLastKeyIndex we use sort.Search which relies on the keys are sorted. If the keys are not sorted, I'm afraid "firstKeyIdx > lastKeyIdx" will not always fail. Do you think we can add a quick check about m.keys[0] is less than m.keys[len(m.keys)-1]?

Comment (tangenta) [reply br/pkg/lightning/backend/external/engine.go:112]: > I'm afraid "firstKeyIdx > lastKeyIdx" will not always fail

OK. I think current implementation is fine enough.

**Review by tangenta — APPROVED** [review]

**Review by wjhuang2016 — APPROVED** [review]

**Pull request merged**